src/vision/harmonic/portable_generic_code/read_and_write.md

```diff
@@ -0,0 +1,155 @@
+    ///                     continue
+    ///                 }
+    ///                 Err(e) => {
+    ///                     return Err(e.into());
+    ///                 }
+    ///             }
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    async fn ready(&mut self, interest: Interest) -> io::Result<Ready>;
+}
+
+pub trait AsyncRead: AsyncIo {
+    fn try_read(&mut self, buf: &mut ReadBuf<'_>) -> io::Result<()>;
+}
+
+pub trait AsyncWrite: AsyncIo {
+    fn try_write(&mut self, buf: &[u8]) -> io::Result<usize>;
+}
+````
+
+This allows users to:
+
+- Take `T: AsyncRead`, `T: AsyncWrite`, or `T: AsyncRead + AsyncWrite`
+
+Note that it is always possible to ask whether writes are "ready", even for a read-only source; the answer will just be "no" (or perhaps an error).
+
+### Can we convert all existing code to this form?
+
+The `try_read` and `try_write` methods are basically identical to the existing "poll" methods. So the real question is what it takes to implement the `ready` async function. Note that tokio internally already adopts a model very similar to this on many types (though there is no trait for it).
+
+It seems like the torture case to validate this is openssl.
+
+## Variant B: Some form of split
+
+Another alternative is to have read/write traits and a way to "split" a single object into separate read/write traits:
+
+```rust
+#[repr(inline_async)]
+trait AsyncRead {
+    async fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize>;
+}
+
+#[repr(inline_async)]
+trait AsyncWrite {
+    async fn write(&mut self, buf: &[u8]) -> std::io::Result<usize>;
+}
+
+#[repr(inline_async)]
+trait AsyncBidirectional: AsyncRead + AsyncWrite {
+    async fn split(&mut self) -> (impl AsyncRead + '_, impl AsyncWrite + '_)
+}
+```
+
+The challenge here is to figure out exactly how that definition should look. The version I gave above includes the possibility that the resulting readers/writers have access to the fields of `self`.
+
+## Variant C: Extend traits to permit expressing that functions can both execute
+
+Ranging further out into unknowns, it is possible to imagine extending traits with a way to declare that two `&mut self` methods could both be invoked concurrently. This would be generally useful but would be a fundamental extension to the trait system for which we don't really have any existing design. There is a further complication that the `read` and `write` methods are in distinct traits (`AsyncRead` and `AsyncWrite`, respectively) and hence cannot
+
+```rust
+#[repr(inline_async)]
+trait AsyncRead {
+    async fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize>;
+    async fn write(&mut self, buf: &[u8]) -> std::io::Result<usize>;
+}
+
+#[repr(inline_async)]
+trait AsyncWrite {
+}
+
+#[repr(inline_async)]
+trait AsyncBidirectional: AsyncRead + AsyncWrite {
+    async fn split(&mut self) -> (impl AsyncRead + '_, impl AsyncWrite + '_)
+}
+```
```

Comment:
A "variant D" to consider here would be to implement `(Async)Read + (Async)Write for &T` like `std` ([example](https://doc.rust-lang.org/std/fs/struct.File.html#impl-Read-1)) and `async_std` ([example](https://docs.rs/async-std/1.9.0/async_std/fs/struct.File.html#impl-Read-1)) do. Being able to split into a read and write pair is then as simple as obtaining a shared handle:

```rust
// Using borrows, being entirely zero-cost
let (reader, writer) = &mut (&stream, &stream);
io::copy(reader, writer).await?;

// Creating owned values, using an Arc which implements
// `impl<T> AsyncRead for Arc<T> where &T: AsyncRead {}`
// such as `async_dup::Arc`
let stream = Arc::new(stream);
io::copy(stream.clone(), stream).await?;
```
Because this is a pattern which already exists in the stdlib and has proven to work in async runtimes, I think it's important to include in conversations on how to solve this.

Reply:
Agreed. That will, however, only work for certain types, right? Not in the general case?

Reply:
I added a placeholder here, I'd like to elaborate on this a bit more.

Reply:
> That will, however, only work for certain types, right? Not in the general case?

It depends on how one defines "the general case". But we've found this to work for all OS primitives which implement `(Async)Read` and `(Async)Write` (including TCP, unix sockets, and files). And crates such as [`async-h1`](https://docs.rs/async-h1/2.3.2/async_h1/) show protocols can be built on top of this [1].

But yeah, for sure happy to elaborate on this more.

---

[1]: `async_h1::accept` takes `AsyncRead + AsyncWrite + Clone`, which can be any of `TcpStream`, `UnixStream`, or `File` wrapped in an `Arc`. This allows us to work around the lack of async lending iterators in the language right now, at the cost of keeping a refcount.

Reply:
Also probably worth adding: a key reason _why_ `(Async){Read,Write}` exists on references, is because we (for the most part) cannot guarantee that only a single handle to a file descriptor is open on the system. This is the same rationale why for example [`TcpStream::set_ttl`](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_ttl) takes `&self` rather than `&mut self`.

This also means that `Arc<Mutex<TcpStream>>` and  `Arc<TcpStream>` provide the same correctness guarantees [1], and why `Read + Write for &TcpStream` is as valid as `Read + Write for TcpStream`.

---

[1]: The emphasis is very much on "guarantees" here.